### DISCUSSION TITLE
Does Pyright support non-conventional import statements which involves a function (e.g. `import_python("foo.bar", "ExampleFunc")`)

### DISCUSSION THREAD

**Discussion by dluo** [microsoft/pyright#9396] [Q&A] [answered]
I am currently working on a codebase which uses python-like syntax but is not truly Python. This involves non conventional import statements such as `import_python("foo.bar", "ExampleFunc")` instead of simply `from foo.bar import ExampleFunc`. A real example of this would be usages of `importlib.__import__`.

I was wondering if this is something that's possible for Pyright to support through one of the configs. My thinking is that if Pyright had to support similar function-like imports in the past, then this convention may already be handle and that there may be existing hooks which I can reuse for my specific use case. For my use case, I only care about string literals, and don't need to support any string interpolation .

**Comment by erictraut** [comment, marked as answer]
Pyright is a [standards-compliant](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) Python type checker. What you're suggesting would be a non-standard extension.

Pyright has no hard-coded knowledge of third-party libraries. All of its knowledge about the semantics of third-party libraries comes from the type annotations provided in the library code or in stub files. There's currently no way in the [Python type system](https://typing.readthedocs.io/en/latest/spec/index.html) to indicate that a library is providing custom import behavior.

If you want to explore something like this, you could propose it in the [Python typing forum](https://discuss.python.org/c/typing/32), but my guess is that there won't be much appetite within the typing community to support non-standard mechanisms like this.

Pyright is open-sourced, so one option is to create your own fork.

Another option is to use a `TYPE_CHECKING` conditional such that pyright sees a conventional `import` statement but the code executed at runtime uses a custom function call.

**Reply by dluo** [reply]
Thanks for the quick response! This was what I expected as well when I looked through the source code, but wanted to double check here in case there was something that I missed. I'll look into the option of creating my own fork to add support for this.
